### ISSUE TITLE
[Bug]: The “backgroundColor” does not take effect immediately when the window is displayed

### ISSUE THREAD

**Issue by classfang** [electron/electron#42523]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

31.0.1

### What operating system are you using?

macOS

### Operating System Version

14.5 (23F79)

### What arch are you using?

arm64 (including Apple Silicon)

### Expected Behavior

```
  const mainWindow = new BrowserWindow({
    width: 900,
    height: 670,
    show: false,
    autoHideMenuBar: true,
    backgroundColor: '#141414',
    ...(process.platform === 'linux' ? { icon } : {}),
    webPreferences: {
      preload: join(__dirname, '../preload/index.js'),
      sandbox: false
    }
  })
```

After creating the window, the background color should be “#141414”.

### Actual Behavior

The moment the window is displayed, the background color is white and immediately changes to “#141414”.

**Comment by classfang** [comment]
**Test Repro Url**
https://github.com/classfang/electron-app

**Comment by hopejr** [comment]
I might add here that this seems to happen even when delaying showing the window after the ready-to-show event has been fired. Whenever the window is shown, it flashes white first, even if the page in it has been loaded.

**Comment by hopejr** [comment]
I've just tested on Windows 11 23H2, and it happens there. So it's not isolated to macOS.

**Comment by hovancik** [comment]
Linux as well.

**Comment by Lwenz95** [comment]
Had the same issue a while ago. For me this fixed it:

```
const createWindow = () => {
    console.log("Creating window...");
    win = new BrowserWindow({
        show: false,
        titleBarStyle: "hidden",
        titleBarOverlay: true,
        webPreferences: {
            preload: path.join(__dirname, "preload.js"),
            contextIsolation: false,
            enableRemoteModule: true,
            nodeIntegration: true,
        }
    });

    attachTitlebarToWindow(win);

    win.once("ready-to-show", () => {
        console.log("Window is ready to show");
        win.maximize();
        win.show();
    });

    win.loadFile("index.html").then(() => {
        console.log("index.html loaded successfully");
    }).catch(err => {
        console.error("Failed to load index.html:", err);
    });
}

try {
    require("electron-reloader")(module);
} catch (err) {
    console.error("Failed to load electron-reloader:", err);
}

app.on("ready", () => {
    console.log("App is ready");
    createWindow();
});
```

It will only show when it's ready. When I started it up it wasn't loaded fully yet.
I set it to show: false so it's not showing when the window is being created.

**Comment by hopejr** [comment]
I also defer showing the window until after the ready-to-show event has fired. It still flashes white for me.
Even putting a delay after the event had fired with setTimeout still has a white flash. It appears that the process of showing the window causes it, even if the page has fully loaded.
This problem may have started as far back as Electron 29, though I see if only on the very rare occasion on that version.

**Comment by Lwenz95** [comment]
Is it related to the color at all? Does it also happen when you put it into a css file or within your html file?

**Comment by hopejr** [comment]
Disabling hardware acceleration doesn't fix it.

I've tried applying the background in the following ways:

- Using the options on the browser window
- Via CSS
- Setting a background-color style on the body tag

Same thing happens on each.

It still happens even after delaying showing the window until 5 seconds after the ready-to-show event is fired.

The easiest way to describe what is happening visually, is the window appears completely white for a fraction of a second, then the content appears fully loaded with the correct background (or vibrancy on macOS) applied. Yes, then even happens with vibrancy set without a background colour on macOS.

**Comment by vojtabohm** [comment]
I am also seeing this issue. It only happens upon the first initial show. It does not happen upon subsequent shows, where the background color is respected). The window is completely empty, not even native controls are shown.

![Elmedia Player - 2024-07-26 at 16 12 30@2x](https://github.com/user-attachments/assets/40024a6d-57fc-4478-ad7b-5d079fadcec8)
![Elmedia Player - 2024-07-26 at 16 12 24@2x](https://github.com/user-attachments/assets/08c9dc23-3ac6-4430-92f5-212b7107741b)

I went into detail in a [SO question](https://stackoverflow.com/posts/78798544).

**Comment by andmm** [comment]
This happens on windows as well on the latest electron version.

**Comment by hopejr** [comment]
Seeing this is affecting more than macOS and multiple people have confirmed it, meaning it's an actual problem, can this issue be tagged appropriately and prioritised? I'm sticking with v29 until this issue is sorted out, which isn't an ideal situation. It's a show-stopper for me.

**Comment by KevinHughes** [comment]
My (macOS) Electron application shows its windows with a "fade in" effect - they start as transparent and then fade in their contents. But this bug makes every window show a flash of white when doing this, which is preventing me from upgrading from v30 to v31.

**Comment by andmm** [comment]
This happens in the latest VS Code version as well.

**Comment by kkocdko** [comment]
@hopejr @andmm  It's unbelivable, on my machine the latest vscode's electron 30.4.0 and electron 33 and electron 29, all have this issue. But in my machine the vscode doesn't effected by this. Turn it to dark mode and restart it, vscode NEVER flicker a white background. I'm curious what magics vscode have.

**Comment by kkocdko** [comment]
I know why VSCode in windows doesn't have this problem. It use  `{ titleBarStyle: "hidden", titleBarOverlay: true }`. And the frameless window with controls will not show white background.

**Comment by kkocdko** [comment]
This works on linux, but not on windows.

```js
app.whenReady().then(async () => {
  await new Promise((resolve) => setTimeout(resolve, 0)); // workaround to reduce flicker in linux
  createWindow();
  app.on("activate", () => {
    if (BrowserWindow.getAllWindows().length === 0) {
      createWindow();
    }
  });
});
```

**Comment by hovancik** [comment]
bump

**Comment by KevinHughes** [comment]
After upgrading to electron 35.0.0, a "white flash" before first render is happening consistently for me again on macOS Sequoia 15.3.1, despite calling `window.show() ` only after the `ready-to-show` event with a one-window app. A basic testcase gist I made could not reproduce the problem, which makes me think there's some specific combination of app, window, and webPreferences options (many of which I use in production in my app) that is triggering this. If I figure it out I'll make a gist.

**Comment by KevinHughes** [comment]
It looks like this was fixed in https://github.com/electron/electron/pull/46353! [clavin](https://github.com/clavin) is a total hero.

**Comment by hopejr** [comment]
Does this fix the issue in Windows too? Looks like the code changes are only for transparent windows in macOS.

**Comment by codebytere** [comment]
@hopejr no, but i'm working on it via https://github.com/electron/electron/issues/46924

**Comment by xmedeko** [comment]
Is still `ready-to-show` necessary to avoid some "flash" after this fix as mentioned in [doc for `ready-to-show`](https://www.electronjs.org/docs/latest/api/browser-window#using-the-ready-to-show-event): "Showing the window after this event will have no visual flash..." Or is the doc obsolete? I do not have flash when using `show: true` flag.
